### ISSUE TITLE
[Windows] Revisit how the RTL handles multi-module apps

### ISSUE THREAD

**Issue by ramosian-glider** [google/sanitizers#332]
Originally reported on Google Code with ID 332

```
We should probably revisit how the RTL handles multi-module apps.

Currently, the main module (.EXE) statically links in the RTL and all the other modules
(.DLLs, basically) statically link with a simple thunk .LIB that delegates all the
ASan interface calls to the RTL in the main module.
[See issue 209 for more background]

This works mostly well in practice, but makes deployment slightly more complex.
Namely, one has to use clang as a linker for DLLs or pass the path to the thunk .LIB
to the link.exe invocations.
COFF .OBJ files have a directive that allows us to specify implicit .LIB dependencies,
but unfortunately the same .OBJ file might be used in an .EXE and a .DLL, thus we can't
use the directive with the current structure of the RTL.

Reid has suggested me to restructure the RTL like this:
- put the RTL into a .DLL
- all the modules (.EXE, .DLL, etc) should statically link with a thunk .LIB that forwards
all the interface calls to the RTL DLL.
- put a directive to implicitly link the thunk .LIB into all the address-sanitized
.OBJ files

This sounds mostly good, but I wonder how would it affect usability?
Basically, which process should put the ASan RTL DLL next to the .EXE?
Is there a directive for that?
```

Reported by `timurrrr@google.com` on 2014-08-08 13:58:38

**Comment by rnk** [comment]
I would still really like to do this, it would drastically simplify deployment if users just have one .lib file to link against instead of one for exes and one for dlls.

This would make the ASan RTL into a DLL. Every DLL comes with a .lib import library, and that import library can actually contain real code. We should be able to jam the interceptors into the import library so that users only have one linker flag to think about.
